File: ДЗ 2/ch-4-2-step-6.swift
class Node {
    var value: Character?
    lazy var left: Node = Node()
    lazy var right: Node = Node()

    func insert(value: Character, with iterator: inout String.Iterator) {
        let char = iterator.next()
        if char == nil {
            self.value = value
        } else {
            if char! == "0" {
                self.left.insert(value: value, with: &iterator)
            } else {
                self.right.insert(value: value, with: &iterator)
            }
        }
    }

    func traverseUntilFound(iterator: inout String.Iterator) -> Character? {
        if self.value != nil {
            return self.value!
        } else {
            let char = iterator.next()
            guard char != nil else {
                return nil
            }

            if char! == "0" {
                return self.left.traverseUntilFound(iterator: &iterator)
            } else {
                return self.right.traverseUntilFound(iterator: &iterator)
            }
        }
    }
}

let count = Int(readLine()!.split(separator: " ")[0])!

var root = Node()
for _ in 0..<count {
    var iterator = readLine()!.makeIterator()
    let char = iterator.next()!
    _ = iterator.next() // skip ":"
    _ = iterator.next() // skip " "
    root.insert(value: char, with: &iterator)
}

var iterator = readLine()!.makeIterator()
var output = ""

while true {
    let char = root.traverseUntilFound(iterator: &iterator)

    if char != nil {
        output.append(char!)
    } else {
        break
    }
}

print(output)

File: ДЗ 2/ch-4-2-step-6.py
class Node:
    def __init__(self):
        self.value = None
        self.left = None
        self.right = None

    def insert(self, value, iterator):
        char = next(iterator, None)
        if char is None:
            self.value = value
        else:
            if char == "0":
                if self.left is None:
                    self.left = Node()
                self.left.insert(value, iterator)
            else:
                if self.right is None:
                    self.right = Node()
                self.right.insert(value, iterator)

    def traverseUntilFound(self, iterator):
        if self.value is not None:
            return self.value
        else:
            char = next(iterator, None)
            if char is None:
                return None

            if char == "0":
                return self.left.traverseUntilFound(iterator)
            else:
                return self.right.traverseUntilFound(iterator)


count = int(input().split()[0])

root = Node()
for _ in range(count):
    line = input()
    char = line[0]
    iterator = iter(line[2:])
    root.insert(char, iterator)

iterator = iter(input())
output = ""
while True:
    char = root.traverseUntilFound(iterator)
    if char is not None:
        output += char
    else:
        break

print(output)
